bin/sidecli.re

```diff
@@ -0,0 +1,243 @@
+  read_identity_file("0/identity.json"),
+  // read_identity_file("1/identity.json"),
+  read_identity_file("2/identity.json"),
+  // read_identity_file("identity_3.json"),
+];
+
+let make_filename_from_address = wallet_addr_str => {
+  String.concat("", [wallet_addr_str, ".tzsidewallet"]);
+};
+
+module Serializable = {
+  [@deriving yojson]
+  type wallet_file = {
+    address: Wallet.t,
+    priv_key: Address.key,
+  };
+};
+
+let exits =
+  Term.default_exits
+  @ Term.[exit_info(1, ~doc="expected failure (might not be a bug)")];
+
+// Commands
+// ========
+
+// Create wallet
+
+let info_create_wallet = {
+  let doc = "Creates a wallet file. The wallet file's filename is its address. The wallet file contains the private key corresponding to that address.";
+  let man = [
+    `S(Manpage.s_bugs),
+    `P("Email bug reports to <contact@marigold.dev>."),
+  ];
+  Term.info("create-wallet", ~version="%‌%VERSION%%", ~doc, ~exits, ~man);
+};
+
+let create_wallet = () => {
+  let (key, wallet) = Wallet.make_wallet();
+
+  let swallet: Serializable.wallet_file = {priv_key: key, address: wallet};
+  let to_yojson = Serializable.wallet_file_to_yojson;
+  let wallet_json = swallet |> to_yojson;
+
+  let wallet_addr_str = Wallet.address_to_string(wallet);
+
+  let filename = make_filename_from_address(wallet_addr_str);
+
+  Yojson.Safe.to_file(filename, wallet_json);
+
+  Lwt.return(`Ok());
+};
+
+let create_wallet_t =
+  Term.(ret(const(() => create_wallet() |> Lwt_main.run) $ const()));
+
+// Transactions
+
+let info_create_transaction = {
+  let doc =
+    Printf.sprintf(
+      "Submits a transaction to the sidechain. The transaction will be communicated to all known validators to be included in the next block. If the path to the wallet file corresponding to the sending address is not provided, a wallet file with the correct filename (%s) must be present in the current working directory",
+      make_filename_from_address("address"),
+    );
+  let man = [
+    `S(Manpage.s_bugs),
+    `P("Email bug reports to <contact@marigold.dev>."),
+  ];
+  Term.info(
+    "create-transaction",
+    ~version="%‌%VERSION%%",
+    ~doc,
+    ~exits,
+    ~man,
+  );
+};
+
+let create_transaction = (address_sender_str, address_receiver_str, amount) =>
+  switch (
+    address_sender_str == "",
+    address_receiver_str == "",
+    amount <= 0,
+    address_receiver_str
+    |> BLAKE2B.of_string
+    |> Option.map(Wallet.address_of_blake),
+  ) {
+  | (true, _, _, _) =>
+    Lwt.return(
+      `Error((
+        false,
+        "Must provide a sender wallet address, or a path to a wallet file.",
+      )),
+    )
+  | (_, true, _, _) =>
+    Lwt.return(`Error((false, "Must provide a destination wallet address.")))
+  | (_, _, true, _) =>
+    Lwt.return(
+      `Error((false, "Must provide a transaction amount above zero.")),
+    )
+  | (_, _, _, None) =>
+    Lwt.return(
+      `Error((false, "Transaction destination must be a valid address.")),
+    )
```

Comment:
Really nice to have nice error messages
![cute anime girl happy with good error messages](https://user-images.githubusercontent.com/3393115/122487746-57da8880-cfb2-11eb-9391-cc4cf50b6cb2.gif)